extensions/native/guest-macro/src/transportation.rs

```diff
@@ -0,0 +1,321 @@
+    pc_diff += jal_example_directives.len() - 1;
+
+    jal_instruction.operands[1] = Operand::usize(DEFAULT_PC_STEP as usize * (pc_diff + 1));
+    add_directives(instruction_to_directives(jal_instruction));
+
+    add_directives(vec![
+        u32_to_directive(GAP_INDICATOR),
+        u32_to_directive(pc_diff as u32),
+    ]);
+
+    for input_var in input_vars {
+        result.push_str(&format!("\t\t\t{} = in(reg) {},\n", input_var, input_var));
+    }
+    for output_var in output_vars {
+        result.push_str(&format!(
+            "\t\t\t{} = out(reg) {},\n",
+            output_var, output_var
+        ));
+    }
+    result.push_str("\t\t)\n");
+    result.push_str("\t}\n");
+    result
+}
+
+pub fn transport_rust_to_edsl<F: Field>(
+    rust_type: String,
+    edsl_type: String,
+    rust_name: String,
+    edsl_fp: usize,
+) -> Vec<MacroInstruction<F>> {
+    match (rust_type.as_str(), edsl_type.as_str()) {
+        ("usize", "Felt<F>") => transport_usize_to_felt(rust_name, edsl_fp),
+        _ => panic!(
+            "Unsupported conversion from rust type {:?} to edsl type {:?}",
+            rust_type, edsl_type
+        ),
+    }
+}
+
+fn transport_usize_to_felt<F: Field>(
+    rust_name: String,
+    edsl_fp: usize,
+) -> Vec<MacroInstruction<F>> {
+    let mut result = vec![];
+    for i in (0..RV32_REGISTER_NUM_LIMBS).rev() {
+        // add [{rust_name} + i] to [edsl_fp]
+        result.push(MacroInstruction::new(
+            VmOpcode::with_default_offset(FieldArithmeticOpcode::ADD),
+            [
+                Operand::usize(edsl_fp),
+                Operand::usize(if i == RV32_REGISTER_NUM_LIMBS - 1 {
+                    0
+                } else {
+                    edsl_fp
+                }),
+                Variable(rust_name.clone(), i),
+                Operand::usize(NATIVE_KERNEL_AS as usize),
+                Operand::u32(if i == RV32_REGISTER_NUM_LIMBS - 1 {
+                    RV32_IMM_AS
+                } else {
+                    NATIVE_KERNEL_AS
+                }),
+                Operand::u32(RV32_REGISTER_AS),
+            ],
+        ));
+        if i > 0 {
+            result.push(MacroInstruction::new(
+                VmOpcode::with_default_offset(FieldArithmeticOpcode::MUL),
+                [
+                    Operand::usize(edsl_fp),
+                    Operand::usize(edsl_fp),
+                    Operand::usize(1 << RV32_CELL_BITS),
+                    Operand::u32(NATIVE_KERNEL_AS),
+                    Operand::u32(NATIVE_KERNEL_AS),
+                    Operand::u32(RV32_IMM_AS),
+                ],
+            ));
+        }
+    }
+    result
+}
+
+pub fn transport_edsl_to_rust<F: Field>(
+    rust_type: String,
+    edsl_type: String,
+    rust_name: String,
+    edsl_fp: usize,
+) -> Vec<MacroInstruction<F>> {
+    match (rust_type.as_str(), edsl_type.as_str()) {
+        ("usize", "Felt < F >") => transport_felt_to_usize(rust_name, edsl_fp),
```

Comment:
i am going to develop some examples tomorrow where a const size array of Felts is converted to array of u32s.  That is needed in my case (verify root proof inside a native-rv32 hybrid program).